### ISSUE TITLE
Is Closed Window Queue really needed?

### ISSUE THREAD

**Issue by kewinrausch** [IRATI/stack#446]
<a href="https://github.com/vmaffione"><img src="https://avatars.githubusercontent.com/u/4920001?v=3" align="left" width="96" height="96" hspace="10"></img></a> **Issue by [vmaffione](https://github.com/vmaffione)**
_Monday Feb 02, 2015 at 10:04 GMT_
_Originally opened as https://github.com/dana-i2cat/irati/issues/443_

---

The current prototype has support for blocking write, meaning that if an IPCP - let's say a normal IPCP - cannot fulfil a write request for a certain flow, the application will block until some room is made available.
If the flow supports flow control, the condition for blocking is "sender window closed AND closed window queue full".
When the windows reopen (e.g. you receive an ACK), some (or all) PDUs from the CWQ are sent (and of course removed from the CWQ) ==> More room is available for writes! We can wake up blocked applications that will not block, either sending new PDUs through the open window or pushing them in the CWQ.

But then my impression is that CWQ is useless.. It's introducing more buffering (and complexities) without benefits. If you didn't have the CWQ, a writing application would block as soon as the sender window is closed, and is waken up as soon as the sender window (snd RWE) is updated (e.g. ACK received).

In other words, I think that the high level concept that is behind the "CWQ" defined in the spec can be simply implemented by the "blocking write" behaviour, without the necessity of a real CWQ queue.

This fact could also become clearer when delimiting will be in place, where buffering will happen there to merge more little SDUs into a few larger PDUs, therefore provinding a buffering service similar to the TCP socket buffer.

**Comment by kewinrausch** [comment]
<a href="https://github.com/edugrasa"><img src="https://avatars.githubusercontent.com/u/325005?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [edugrasa](https://github.com/edugrasa)**
_Tuesday Feb 03, 2015 at 09:47 GMT_

----

In my opinion it is needed, it greatly influences the goodput experienced by the application (for example, closed window queue size = 10 goodput = 110 Mbps; closed window queue size = 200 goodput = 850 Mbps, of course with our current not very sophisticated flow control policies). Since this is really a spec discussion, you should bring this issue to irati-tech or, even better, psoc-specs :)

**Comment by edugrasa** [comment]
@vmaffione this issue should be closed

**Comment by sandervrijders** [comment]
Shouldn't this be discussed on the ML first? Isn't the CWQ just an extra buffer?

**Comment by edugrasa** [comment]
It's part of the EFCP spec, this should be discussed at the spec level first. It can be configured to whatever size you want, its effect is to smooth the opening and closing of the transmission window as seen by the application (so it prevents unneeded blocking sometimes and enhances throughput). 

So the procedure should be to first discuss this with John and others in the WP3 mailing list, reach and agreement, change the EFCP spec in case the conclusions is that the closed window queues is not needed and then update the implementation accordingly.

Meanwhile, this issue should be closed, because the closed window queue is part of the current EFCP specification.

**Comment by sandervrijders** [comment]
@vmaffione Care to start this discussion?

**Comment by edugrasa** [comment]
This is a specs discussion. Closed window queue is at the EFCP spec, so please discuss this at the wp2 pristine mailing list, or even better, psoc-specs mailing list.

**Comment by sandervrijders** [comment]
Yes, eactly, that's what I meant. Start the discussion on the psoc ML.

**Comment by edugrasa** [comment]
Ok, just have good arguments on why you think it is not needed :)
